### PR TITLE
ascanrules: Fix xss context false negative

### PR DESCRIPTION
**Pull request by tmendo**
- Fixes an XSS false negative when increasing the threshold to high
- Move HtmlContext and HtmlContextAnalyser from core to ascanrules
- Exclude injected tag from being considered part of the context, which was causing the strict context comparison to fail
- Fix search for XSS outside of any tag requiring the parentTag to exist.

Closes https://github.com/zaproxy/zaproxy/issues/6727

### CODE REVIEW

**Comment by thc202** [comment]
Some comments:
 - Needs tests (both for the scan rule and the `HtmlContextAnalyser`);
 - Changelog needs to be updated.

**Comment by kingthorin** [comment]
If there were changes to these files while moving them could you summarize or provide a diff?

```diff
- import org.zaproxy.zap.httputils.HtmlContext;
- import org.zaproxy.zap.httputils.HtmlContextAnalyser;
+ import org.zaproxy.zap.extension.ascanrules.utils.HtmlContext;
+ import org.zaproxy.zap.extension.ascanrules.utils.HtmlContextAnalyser;
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java [hunk 1]
```diff
@@ -687,6 +687,7 @@ protected Response serve(IHTTPSession session) {
         HttpMessage msg = this.getHttpMessage(test + "?name=test");
 
         this.rule.init(msg, this.parent);
+        this.rule.setAlertThreshold(AlertThreshold.HIGH);
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java:690]: Should be in a new test.

Comment (tmendo) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java:690]: ok, will do

**Comment by tmendo** [comment]
> If there were changes to these files while moving them could you summarize or provide a diff?
> 
> ```diff
> - import org.zaproxy.zap.httputils.HtmlContext;
> - import org.zaproxy.zap.httputils.HtmlContextAnalyser;
> + import org.zaproxy.zap.extension.ascanrules.utils.HtmlContext;
> + import org.zaproxy.zap.extension.ascanrules.utils.HtmlContextAnalyser;
> ```

the PR has two commits: the 1st is just the move from the core, without changes except for fixing the import path. Later we will do a PR to remove those from the core.

ps: we == me and @DiogoMRSilva

**Comment by thc202** [comment]
They should not be removed yet just deprecated.

**Comment by kingthorin** [comment]
Thanks, I was looking at overall changes. You're right it's obvious that all that really changed in the second commit was the addition of a conditional in identifying the start position. Thanks for pointing out the two commits 👍

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/utils/HtmlContextAnalyser.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.network.HttpMessage;
+
+public class HtmlContextAnalyser {
+
+    private char[] quotes = {'\'', '"'};
+
+    // Tag attributes which can contain javascript
+    private String[] scriptAttributes = {
+            "onBlur",
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/utils/HtmlContextAnalyser.java]: While moving from core some indentation changes were done, better to amend the commit to keep the previous indentation (similar comment for the other class, has a the moved import).

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java (hunk 1 above)
Comment (tmendo) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java:690]: hum, I can duplicate the test with the threshold High, but does it make sense? As is, we are basically saying the XSS is found even after raising the bar to High.
With two tests we have the advantage of ensuring the expected payload is used in both scenarios.
So just want to confirm the addition @thc202, before pushing.

Comment (kingthorin) [reply addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/CrossSiteScriptingScanRuleUnitTest.java:690]: You can turn this into a parameterized test. But a test should be added that fails before your change (doesn't identify the vuln) but does after your changes.

**Comment by tmendo** [comment]
Apparently, I missed some small, but significant, changes on the XSS rule and the original problem was fixed on [06ced036bb50fabdf1f6601092f5025604cce559](https://github.com/zaproxy/zap-extensions/commit/06ced036bb50fabdf1f6601092f5025604cce559) by adding more payloads. 

The context extraction (the payload tag being considered part of the tag hierarchy) is not necessarily wrong since we can disable exact context matching, as it is currently done. I will get up to date with the changes and submit other PRs as necessary. Thanks for your time!
